src/lib/operators/sort.cpp

```diff
@@ -200,41 +244,69 @@ class Sort::SortImpl : public AbstractReadOnlyOperatorImpl {
-  void _materialize_sort_column() {
+  void _materialize_sort_column(const std::shared_ptr<PosList>& pos_list = nullptr) {
     auto& row_id_value_vector = *_row_id_value_vector;
     row_id_value_vector.reserve(_table_in->row_count());
 
     auto& null_value_rows = *_null_value_rows;
 
-    const auto chunk_count = _table_in->chunk_count();
-    for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
-      const auto chunk = _table_in->get_chunk(chunk_id);
-      Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
+    if (pos_list) {
+      auto segment_ptr_and_accessor_by_chunk_id =
+          std::unordered_map<ChunkID, std::pair<std::shared_ptr<const BaseSegment>,
+                                                std::shared_ptr<AbstractSegmentAccessor<SortColumnType>>>>();
+      segment_ptr_and_accessor_by_chunk_id.reserve(_table_in->chunk_count());
 
-      auto base_segment = chunk->get_segment(_column_id);
+      // When there was a preceding sorting run, we materialize according to the passed PosList.
+      for (RowID row_id : *pos_list) {
+        const auto [chunk_id, chunk_offset] = row_id;
 
-      segment_iterate<SortColumnType>(*base_segment, [&](const auto& position) {
-        if (position.is_null()) {
-          null_value_rows.emplace_back(RowID{chunk_id, position.chunk_offset()}, SortColumnType{});
+        auto& segment_ptr_and_typed_ptr_pair = segment_ptr_and_accessor_by_chunk_id[chunk_id];
+        auto& base_segment = segment_ptr_and_typed_ptr_pair.first;
+        auto& accessor = segment_ptr_and_typed_ptr_pair.second;
+
+        if (!base_segment) {
+          base_segment = _table_in->get_chunk(chunk_id)->get_segment(_column_id);
+          accessor = create_segment_accessor<SortColumnType>(base_segment);
+        }
+
+        // If the input segment is not a ReferenceSegment, we can take a fast(er) path
+        if (accessor) {
+          const auto typed_value = accessor->access(chunk_offset);
+          if (!typed_value) {
+            null_value_rows.emplace_back(row_id, SortColumnType{});
+          } else {
+            row_id_value_vector.emplace_back(row_id, typed_value.value());
+          }
         } else {
-          row_id_value_vector.emplace_back(RowID{chunk_id, position.chunk_offset()}, position.value());
+          const auto value = (*base_segment)[chunk_offset];
+          if (variant_is_null(value)) {
+            null_value_rows.emplace_back(row_id, SortColumnType{});
+          } else {
+            row_id_value_vector.emplace_back(row_id, boost::get<SortColumnType>(value));
+          }
         }
-      });
+      }
+    } else {
+      const auto chunk_count = _table_in->chunk_count();
+      for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
+        const auto chunk = _table_in->get_chunk(chunk_id);
+        Assert(chunk, "Did not expect deleted chunk here.");  // see #1686
```

Comment:
Maybe add full URL https://github.com/hyrise/hyrise/pull/1686

Reply:
Done in f6ea5af89bdb12244547fb8a313f682459053c31